### DISCUSSION TITLE
Method for Clean up Openssl 3.0.8

### DISCUSSION THREAD

**Discussion by DStarts23** [openssl/openssl#25518] [Q&A] [answered]
I am using openssl 3.0.8 version, and have two questions here:
Q:1:- Does the code for un-initialization for OpenSSL 3 libraries is okay? 
initialized OpenSSL configuration as:

// Initialize OpenSSL with configuration settings
`if (!OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL)) {
	// Handle initialization failure
	Log.Debug("OpenSSL initialization failed.");
}`
And loading non-fips algorithms as:
`OpenSSL_add_all_algorithms();
OpenSSL_add_all_digests();
OpenSSL_add_all_ciphers();`

For un-initialization, I am using these API's :
`EVP_cleanup();
CRYPTO_cleanup_all_ex_data();
ERR_free_strings();
SSL_COMP_free_compression_methods();`

Q:02: - When I load the Fips provider using this API:
`fipsProvider = OSSL_PROVIDER_load(NULL, "fips);`

I got some value in `fipsProvider` variable, and then after doing work, I unload the fipsProvider using this API:
if(fipsProvider){
  OSSL_PROVIDER_unload(fipsProvider);
}

But in call stack, it still shows value in fipsProvider variable. So I used this line after calling unload API:
`fipsProvider = NULL;`

Do you think after calling unload( ) function, I can assign NULL value to fipsProvider? Is it good approach to unload the Fips Provider?

**Comment by nhorman** [comment]
A1) Its a bit hard to say if what you are doing is correct or not without some sample code to look at.  Though generally speaking, you shouldn't need to explicitly unload digests/ciphers/etc.  You can do that with a call to OpenSSL_cleanup before your application exits

A2) yes, once you've unloaded your fips provider, the pointer to the provider you received is no longer valid, and can be set to null

**Comment by DStarts23** [comment]
@nhorman But I have read that using Openssl_Cleanup( ) is not recommended, as once we clean up everything we might not be able initialize the crypto again, unless we need to restart the application.
https://docs.openssl.org/master/man3/OPENSSL_init_crypto/#description

`The OPENSSL_cleanup() function deinitialises OpenSSL (both libcrypto and libssl). All resources allocated by OpenSSL are freed. Typically there should be no need to call this function directly as it is initiated automatically on application exit. This is done via the standard C library atexit() function. In the event that the application will close in a manner that will not call the registered atexit() handlers then the application should call OPENSSL_cleanup() directly. Developers of libraries using OpenSSL are discouraged from calling this function and should instead, typically, rely on auto-deinitialisation. This is to avoid error conditions where both an application and a library it depends on both use OpenSSL, and the library deinitialises it before the application has finished using it.

Once OPENSSL_cleanup() has been called the library cannot be reinitialised. Attempts to call OPENSSL_init_crypto() will fail and an ERR_R_INIT_FAIL error will be added to the error stack. Note that because initialisation has failed OpenSSL error strings will not be available, only an error code. This code can be put through the openssl errstr command line application to produce a human readable error (see [openssl-errstr(1)](https://docs.openssl.org/master/man1/openssl-errstr/)).`

**Reply by nhorman** [reply]
That is all true, butbgs the question: why are you deinitalizig parts of the library if you're not done using.it?

**Reply by DStarts23** [reply]
I have a confusion over my requirement, so that' why I was thinking to call the Openssl_cleanup( ) method is right or wrong. Now I am clear about it. Thanks for your help.

Besides this, I have one question, If I want to load the Fips provider programmatically in my program without using openssl.cnf file, is it possible? If I bypass the openssl.cnf and load the Fips or default provider directly in my program, does it affect other openssl 3 API or functions?

**Reply by t8m** [reply]
You need to have a config file to be able to load the fips provider as the fips provider needs to load some configuration from the fipsmodule.cnf file and that must be included from a config file that is read by OpenSSL. You do not have to have the fips module automatically activated by the configuration file though.

**Reply by DStarts23** [reply]
@t8m I am working on Windows and using the compiled files from the OpenSSL 3.0.8 version, added as External in my code. Instead of creating the System Environment Variable `OPENSS_CONF`, `OPENSSL_MODULE` and assigning them the values of openssl.cnf file and other dll files. I want to programmatically load the openssl.cnf file in my code. 

For this purpose, I use this API:

```
//const char* config_file_path = "path\\to\\mygitproject\\Win32\\Debug\\openssl.cnf";
// Load the specific OpenSSL configuration file
if (CONF_modules_load_file(config_file_path, NULL, 0) <= 0) {
	printf("Failed to load OpenSSL config file: %s\n", config_file_path);
	ERR_print_errors_fp(stderr);
	
}
```

sometimes it loads the file perfectly, while sometimes it shows error. Plus, The content in my openssl.cnf file is:

```
.include  "full/path/to/fipsmodule.cnf"

[openssl_init]
providers = provider_sect

# List of providers to load
[provider_sect]
default = default_sect
fips = fips_sect

[default_sect]
activate = 1
```

The requirement is the application loads the default provider automatically, while the fips provider will load by user choice programmatically. 

Then I also load the fipsmodule.cnf in my file manually:

```
// Ensure the fipsmodule.cnf path is correct
const char* fipsConfigPath = "path\\to\\mygitProject\\Win32\\Debug\\fipsmodule.cnf";

// Initialize settings for OpenSSL
OPENSSL_INIT_SETTINGS* settings = OPENSSL_INIT_new();
if (settings == NULL) {
	Log.Debug("Failed to create OPENSSL_INIT_SETTINGS.");
	return 0;
}

// Set the configuration file for OpenSSL
OPENSSL_INIT_set_config_filename(settings, fipsConfigPath);

// Initialize OpenSSL with FIPS configuration
if (OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, settings) != 1) {
	Log.Debug("Failed to initialize OpenSSL with FIPS configuration.");
	OPENSSL_INIT_free(settings);
	return 0;
}

//Now try loading the FIPS provider
fipsProvider = OSSL_PROVIDER_load(NULL, "fips");
```

When my breakpoint hits this line: `if (CONF_modules_load_file(config_file_path, NULL, 0) <= 0)` it shows in build output:

```
'AppTest.exe' (Win32): Loaded 'path\to\mygitProject\Win32\Debug\fips.dll'. 
'AppTest.exe' (Win32): Unloaded 'path\to\mygitProject\Win32\Debug\fips.dll'
```

means load the Fips and then unloads it at the same time, I clear everything before running the application cache memory and whatever is that so.
and when I further  hits the other breakpoints, the value in fips_Provider remains NULL. Why is that so?

**Reply by t8m** [reply]
1. Is there anything before the `.include` directive in the openssl.cnf file? If not, this is wrong. There must be at least `openssl_conf = openssl_init` line.
2. You cannot load fipsmodule.cnf directly. It must be loaded through inclusion from the openssl.cnf file because that contains the necessary openssl_init and provider_sect sections.
3. To prevent activating the fips provider automatically you need to remove the line `activate = 1` from the fipsmodule.cnf file.

I do not think you can get around setting the OPENSSL_MODULE variable unless you install the openssl (including the fips provider) with `make install`.

**Reply by DStarts23** [reply]
@t8m It's quite weird that I am able to load the Fips and default provider in my application even without installing the OpenSSL 3.x on my machine using this config file to load the openss.cnf file: `CONF_modules_load_file(config_file_path, NULL, 0)` 

Plus, our team had worked on previous OpenSSL version 1.0.2u and they were only adding the dll files or lib files in their application code, not the openssl.cnf files.

I have been asked by few questions:

1. Does it necessary to have `openssl.cnf` file dependency? Can't we work without it as in OpenSSL 1.0.2u, or in openSSL 1.1.1c versions, we didn't use the openssl.cnf file to load anything. Can't providers load without it?

2. If we use the openssl.cnf file, it means do we also need to pass the path of the loaded providers as well? Just like currently I am passing the full path of `fipsmodule.cnf` file in openssl.cnf to load the fips.dll, otherwise by passing the relative path of fipsmodule, it does not load the fips.dll.  

3. Which provider is the best for non-FIPS compliance? Base, Legacy, NULL, Default? or does legacy provider holds the Previous openssl versions algorithms and digest?

4. Does openssl 3.0.8 version compilation results in fips.lib file? Or does it only generate the fips.dll/fips.so file?

The foremost question is related to the `openssl.cnf` file, because we have a huge application that requires to load the FIPS module, which in turn has many services to load individually, that further requires the openssl.cnf separately there including fipsmodule.cnf path.

**Reply by t8m** [reply, marked as answer]
1.0.2 had a completely different way how to ensure FIPS compliance - there was the possibility to build it with so called FIPS container module. No configuration file was needed for that.

1.1.1 does not have FIPS support at all apart from some 3rd party vendor-patched versions, which we do not know much about.

Re 1. Yes, openssl.cnf is absolutely needed for loading the FIPS provider, its contents can be fairly minimal.

Re 2. If the provider is installed into the installation directory as per make install, there does not have to be a full path. Otherwise yes.

Re 3. Normal operation (non-fips-compliant) is provided by the default provider. Base provider must be loaded along the fips provider (not needed with default provider) to be able to decode and encode keys and certificates. Legacy provider supports legacy algorithms.

Re 4. The fips provider must be a loadable module. Static linking of the fips provider is not supported.
